ggplot2 tidy eval post

batpigandme opened the pull request:
* Tidy eval section
* @hadley, I'm realising this is getting long — looking for input re. whether it might be a good idea to give tidy eval in ggplot2 its own post

hadley commented:
I think we should just make this a tidy eval post, and not worry about a developer specific post for now.

batpigandme commented:
OK, refocused to just hit on tidy evaluation. I added subheaders which I'm not sure about. Among other things, we need new CSS to style code that's inside of headers. The font-size for `h3` is 1.6em, but that looks *huge* in the code font. My recommended CSS would be (I'll file in tidytemplate):
```CSS
h3 code {
    font-size: 1em;
}
```

hadley commented on content/articles/2017-07-ggplot2-3-0-0-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,116 @@
+---
+title: ggplot2 3.0.0 - tidy evaluation
```
Maybe just "Tidy evaluation in ggplot2"?

hadley commented on content/articles/2017-07-ggplot2-3-0-0-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,116 @@
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>", fig.retina = 2, dpi = 300)
+library(tidyverse)
+```
+
````
Add in line `<style>` for now?

hadley commented on content/articles/2017-07-ggplot2-3-0-0-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,116 @@
+
+### Facetting with `vars()`
+
+To support quasiquotation in facetting we’ve added a new helper function: [`vars()`](https://ggplot2.tidyverse.org/reference/vars.html), short for variables. Instead of `facet_grid(x + y ~ a + b)` you can now write `facet_grid(vars(x, y), vars(a, b))`. The formula interface won’t go away; but the new `vars()` interface should be much easier to program with.
```
"should be much easier to program with" ->  "supports tidy evaluation, so can be easily programmed with"

hadley commented on content/articles/2017-07-ggplot2-3-0-0-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,116 @@
+p + wrap_by(vs, am)
+```
+
+In our `wrap_by()` function above, we used tidy dots ([`...`](https://adv-r.hadley.nz/quasiquotation.html#dot-dot-dot-...)), which represent an arbitrary number of additional arguments. For a function to use *named arguments*, we'll need to "enquote" the named argument with [`enquo()`](http://rlang.r-lib.org/reference/quotation.html). To create a default name, we'll use [`quo_name()`](http://rlan ...
````
Can you explicit use "quote and unquote" phrase somewhere? That's how we're trying to get people to understand this.

hadley commented on content/articles/2017-07-ggplot2-3-0-0-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,116 @@
+p + wrap_cut(drat)
+```
+
+Symptomatic error message:
````
I think you can remove this now

batpigandme commented:
Just realised that the splicing with named args isn't working as I thought it would…
``` r
library(tidyverse)
mans <- c("chevrolet", "dodge", "ford", "toyota")

year <- 2018

mpg2 <- mpg %>%
  filter(manufacturer %in% mans)

d <- ggplot(mpg2, aes(displ, cty)) + geom_point()

args <- list(rows = quo(year), cols = quo(manufacturer))

d + facet_grid(vars(!!!args))
```

![](https://i.imgur.com/zNApuTg.png)

``` r

d + facet_grid(vars(year), vars(manufacturer))
```

![](https://i.imgur.com/qVWtBfD.png)

Created on 2018-07-13 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

batpigandme commented:
Another @edgararuiz option
```R
easy_cols <- function(df, ...){
  vars <- enexprs(...)
  x <- vars$x
  y <- vars$y
  if(names(vars[1]) == "") x <- vars[[1]]; vars[[1]] <- NULL
  if(names(vars[2]) == "") y <- vars[[2]]; vars[[2]] <- NULL
  ggplot(df) +
    geom_col(aes(x = !!x, y = !!y, !!! vars))
}

easy_cols(mtcars, am, fill = as.factor(am), y = sum(wt))
```

batpigandme commented:
ping @hadley

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+scatter_by(mtcars, disp, drat)
+```
+
+### Facetting with `vars()`
````
Tidy facets with `vars()`?

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+
+### `aes()` contains quosures
+
+The primary developer-facing change is that `aes()` now contains [quosures](https://adv-r.hadley.nz/evaluation.html#quosures) (expression + environment pairs), rather than symbols. As a result, you'll need to take a different approach to extracting the information you need. Below, we use a "quote-and-unquote" pattern for the aesthetic parameter `x_var`.
```
I think this paragraph and example can be moved to the bottom of the post, and replaced with something like this:

## Developer facing changes

If you're building an extension package on top of ggplot2, you need to be aware that the introduction of tidy evaluation fundamentally changes the data structures that `aes()` uses. In brief:

```R
mapping <- aes(mpg, colour = "smoothed")

# Variables are now stored as quosures
mapping$x

# Constants (atomic vectors of length 1), remain
# as is
mapping$colour
```

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+```{r piechart-qq}
+piechart <- function(data, var, ...) {
+  var <- enquo(var)
+  piechart_basic(data, aes(!!factor(1), fill = !!var))
````
Don't need `!!` in front of `factor()`

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+piechart_basic(mpg, aes(factor(1), fill = class))
+```
+
+By quoting and unquoting the arguments, the user can leave the arguments bare when calling the function.
````
The key to calling a tidy evaluation function inside another function is to quote (with `enquo()`) and unquote (with `!!`):

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+The basic pattern is to enquote and unquote named arguments.
+
+```{r scatter-by}
+scatter_by <- function(data, x, y) {
````
We could use this same pattern to make a scatterplot wrapper:

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+
+  ggplot(data) + geom_point(aes(!!x, !!y))
+}
+scatter_by(mtcars, disp, drat)
```
Maybe call functions `pie_chart()` and `scatter_plot()`?

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+}
+</style>
+</html>
+## Tidy evaluation
```
I think we can use h2 for all the headings, and remove the styling fix.

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+
+One of the biggest changes in [ggplot2](https://ggplot2.tidyverse.org/) 3.0.0 is support for [tidy evaluation](https://adv-r.hadley.nz/evaluation.html#tidy-evaluation), making it more programmable, and more consistent with the rest of the tidyverse. Though this introduces some breaking changes, we believe it to be worthwhile in the interest of improving future code. Here, we outline some of the m ...
+
+### `aes()` contains quosures
```
I think we can use h2 for all the headings, and remove the styling fix.

batpigandme replied:
Doesn't h2 need the same styling fix? 

![image](https://user-images.githubusercontent.com/831732/43144202-89e4477e-8f2a-11e8-8674-cb10323e7d3b.png)

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+</html>
+## Tidy evaluation
+
+One of the biggest changes in [ggplot2](https://ggplot2.tidyverse.org/) 3.0.0 is support for [tidy evaluation](https://adv-r.hadley.nz/evaluation.html#tidy-evaluation), making it more programmable, and more consistent with the rest of the tidyverse. Though this introduces some breaking changes, we believe it to be worthwhile in the interest of improving future code. Here, we outline some of the m ...
```
Here, we outline some of the more prominent of those changes. -> Here we show a couple of examples of how you might use it.

"For our complete list of changes ..." -> New developer facing changes section (see below)

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+
+You can now use [quasiquotation](https://adv-r.hadley.nz/quasiquotation.html) in [`aes()`](https://ggplot2.tidyverse.org/reference/aes.html), [`facet_wrap()`](https://ggplot2.tidyverse.org/reference/facet_wrap.html), and [`facet_grid()`](https://ggplot2.tidyverse.org/reference/facet_grid.html). For `aes()`, quasiquotation (`!!`, `!!!`, `:=`) replaces [`aes_()`](https://ggplot2.tidyverse.org/refer ...
+
+Here we'll use quasiquotation in place of `aes_()` to make a function to make a pie chart that allows the user to avoid using quotation marks when they pass their arguments to the function.[^1] The initial function, below, would require that the user know the exact specifications to pass to `aes()`.
```
remove "in place of `aes_()`"

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+p + facet_grid(rows = vars(drv))
+```
+
+Using quosures ensures that the variable comes from the context of the dataframe. Since the ellipsis is the first and only argument of `vars()`, you can use unquote splicing with the `!!!` operator to pass in a list of arguments.
````
Replace para with something like:

Compared to `aes()`, `vars()` takes unnamed arguments. This makes it a more natural fit for use with `!!!`, the unquote-splice operator.

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
```diff
@@ -0,0 +1,150 @@
+  ggplot() +
+    geom_point(aes(displ, cty))
+
+args <- list(quo(year), quo(manufacturer))
```
```
args <- quos(year, manufacturer)
```

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+p + facet_grid(vars(Cylinder = cyl), labeller = label_both)
+```
+
+### Using `vars()` for wrapper functions
````
Remove this heading

hadley commented on content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown:
````diff
@@ -0,0 +1,150 @@
+p + wrap_by(vs, am)
+```
+
+In our `wrap_by()` function above, we used tidy dots ([`...`](https://adv-r.hadley.nz/quasiquotation.html#dot-dot-dot-...)), which represent an arbitrary number of additional arguments. For a function to use *named arguments*, we'll need to quote the named argument with [`enquo()`](http://rlang.r-lib.org/reference/quotation.html). To create a default name, we'll use [`quo_name()`](http://rlang.r- ...
````
For a function to... -> "Alternatively, we could allow the user to provide a single named argument..."

batpigandme commented:
n.b. I forsook everything I know about grammar and let "developer facing changes" exist without a hyphen, since we don't use it elsewhere.

hadley commented:
Looks good. Do you mind tweaking the aspect ratios of the facetting examples so that the individual panels are close to square? i.e. `fig.asp` should be `0.612 * nrow / ncol` (or possibly `0.612 * ncol / nrow`)

batpigandme commented:
> fig.asp should be 0.612 * nrow / ncol
So, for the one with just 3 panels, that's like — there are two more with 8 and 4 panels, respectively…
![image](https://user-images.githubusercontent.com/831732/43160057-172e54ea-8f52-11e8-8dd2-48a0c2e5822d.png)

batpigandme commented:
OK, I modified the aspect ratios, but also changed the `out.width` and `fig.width`s because they got super tall. You can see an example where I _don't_ change the `fig.width`, but do change the `out.width`, because I wasn't sure which way to go…obviously I'll edit them for consistencey.

hadley commented:
Oh hmmm, maybe what I want is for you to keep the height consistent but make the width narrower (to achieve the desired aspect ratio)

batpigandme commented:
OK, so the consistent height for each fig is 432 (with the doc settings `out.width = "700px"`, and `fig.asp = 0.618`).

So, for a 3 panel facet, that means the `fig.asp = "1.836"` (since you suggested 0.612) and `out.width = "218px"
![image](https://user-images.githubusercontent.com/831732/43165978-45234e58-8f63-11e8-86ce-fb3404a9c430.png)

hadley commented:
I think you always have to change both fig.width and out.width to keep the ratio between the width of the image, and the width of the `<img>` constant.

batpigandme commented:
Yeah, I updated that if you look at the preview (sorry, now on phone)

hadley commented:
Ah ok, perfect, thanks!
